pkg/controller/node/nodecontroller.go

```diff
@@ -601,6 +611,80 @@ func (nc *NodeController) Run() {
 	}()
 }
 
+// addOrRemoveNoScheduleTaintByCondition taint or remove NoSchedule taint on this node based on it's condition change.
+func (nc *NodeController) addOrRemoveNoScheduleTaintByCondition(node *v1.Node) (bool, bool) {
+	var changedToUnSchedulable, changedToSchedulable bool
+	observedMap := map[v1.NodeConditionType]*v1.NodeCondition{}
+	_, observedMap[v1.NodeReady] = v1.GetNodeCondition(&node.Status, v1.NodeReady)
+	_, observedMap[v1.NodeOutOfDisk] = v1.GetNodeCondition(&node.Status, v1.NodeOutOfDisk)
+	_, observedMap[v1.NodeNetworkUnavailable] = v1.GetNodeCondition(&node.Status, v1.NodeNetworkUnavailable)
+
+	// If there are saved node status, it only make sense to check conditions when some change happened.
+	if savedNodeStatus, found := nc.nodeStatusMap[node.GetName()]; found {
+		for conditionType, condition := range observedMap {
+			_, savedCondition := v1.GetNodeCondition(&savedNodeStatus.status, conditionType)
+			// something changed
+			if (condition != nil && savedCondition == nil) ||
+				(condition != nil && savedCondition != nil && condition.Status != savedCondition.Status) {
+				if !isSchedulable(condition) {
+					changedToUnSchedulable = true
+					changedToSchedulable = false
+					// if one of the condition changed and becomes unschedulable, exit loop
+					break
+				} else {
+					changedToSchedulable = true
+				}
+			}
+			// if condition == nil && savedCondition != nil, we do nothing
+		}
+	} else {
+		// Otherwise, there are no saved node status, it always worth to check conditions
+		for _, condition := range observedMap {
+			if condition != nil && !isSchedulable(condition) {
+				changedToUnSchedulable = true
+				changedToSchedulable = false
+				break
+			} else {
+				changedToSchedulable = true
+			}
+		}
+	}
```

Comment:
I think here you should just check the current conditions and decide about Unschedulable and Schedulable, instead of comparing with previous saved conditions. My guess is that AddOrUpdateTaintOnNode and RemoveTaintOffNode should be doing implicit checking if there is any existing NoScedule taint or not before adding/removing it.  This would really simplify this code I think.